core/base/dispatch_helper.hpp

```diff
@@ -157,32 +178,54 @@ auto run(T obj, Func&& f, Args&&... args)
  * run uses template to go through the list and select the valid
  * template and run it.
  *
- * @tparam K  the current shared_ptr type tried in the conversion
+ * @tparam Base  the Base class with one template
+ * @tparam ...Types  the types that will be tried with Base, i.e. Base<Types>...
+ * @tparam T  the type of input object waiting converted
+ * @tparam Func  the function will run if the object can be converted to pointer
+ *               of const Base<K>
+ * @tparam ...Args  the additional arguments for the Func
+ *
+ * @param obj  the input object waiting converted
+ * @param f  the function will run if obj can be converted successfully
+ * @param args  the additional arguments for the function
+ *
+ * @return  the result of f invoked with obj cast to the first matching type
+ */
+template <template <class> class Base, typename... Types, typename T,
+          typename Func, typename... Args>
+auto run(T* obj, Func&& f, Args&&... args)
+{
+    return run<Base<Types>...>(obj, std::forward<Func>(f),
+                               std::forward<Args>(args)...);
+}
+
+
+/**
+ * run uses template to go through the list and select the valid
+ * template and run it.
+ *
+ * @tparam K  the current type to try in the conversion
  * @tparam ...Types  the other types will be tried in the conversion if K fails
- * @tparam T  the type of input object
+ * @tparam T  the type of input object waiting converted, has to be either
+ *            unique_ptr or shared_ptr
  * @tparam Func  the function type that is invoked if the object can be
- *               converted to std::shared_ptr<K>
+ *               converted to pointer of const Base<K>
```

Comment:
```suggestion
 *               converted to pointer of Base<K>
```
The const depends on the given type now, right?